src/wallet/wallet.cpp

```diff
@@ -4135,3 +4138,110 @@ bool CWalletTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState&
+{
+    CKeyID keyid = key.GetID();
+    if (key.IsCompressed()) {
+        CTxDestination segwit = WitnessV0KeyHash(keyid);
+        CTxDestination p2sh = CScriptID(GetScriptForDestination(segwit));
+        return std::vector<CTxDestination>{std::move(keyid), std::move(p2sh), std::move(segwit)};
+    } else {
+        return std::vector<CTxDestination>{std::move(keyid)};
+    }
+}
+
+CTxDestination CWallet::AddDestinationForKey(const CPubKey& key, OutputType type)
+{
+    switch (type) {
+    case OUTPUT_TYPE_LEGACY: return key.GetID();
+    case OUTPUT_TYPE_P2SH:
+    case OUTPUT_TYPE_BECH32: {
+        CTxDestination witdest = WitnessV0KeyHash(key.GetID());
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(*this, witprog)) return key.GetID();
+        // Even though CBasicKeyStore knows about P2WPKH redeemscripts implicitly, add it explicitly
+        // so that downgrades to software without this are supported (as long as the wallet file is
+        // up to date).
+        AddCScript(witprog);
+        if (type == OUTPUT_TYPE_P2SH) {
+            return CScriptID(witprog);
+        } else {
+            return witdest;
+        }
+    }
+    default: assert(false);
+    }
+}
+
+CTxDestination CWallet::AddDestinationForScript(const CScript& script, OutputType type)
+{
+    // Note that scripts over 520 bytes are not yet supported.
+    switch (type) {
+    case OUTPUT_TYPE_LEGACY:
+        return CScriptID(script);
+    case OUTPUT_TYPE_P2SH:
+    case OUTPUT_TYPE_BECH32: {
+        WitnessV0ScriptHash hash;
+        CSHA256().Write(script.data(), script.size()).Finalize(hash.begin());
+        CTxDestination witdest = hash;
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(*this, witprog)) return CScriptID(script);
+        // Add the redeemscript, so that P2WSH and P2SH-P2WSH outputs are recognized as ours.
+        AddCScript(witprog);
+        if (type == OUTPUT_TYPE_BECH32) {
+            return witdest;
+        } else {
+            return CScriptID(witprog);
+        }
+    }
+    default: assert(false);
+    }
+}
+
+void CWallet::AddRelatedScripts(const CScript& scriptPubKey, const CKeyID& keyid)
+{
+    CTxDestination segwit = WitnessV0KeyHash(keyid);
+    CScript witprog = GetScriptForDestination(segwit);
+    CTxDestination p2sh = CScriptID(witprog);
+    // For SegWit scriptPubKeys, add the redeemscript explicitly to the wallet, so that
+    // older software versions (without implicit scripts) correctly treat these as ours.
+    if (scriptPubKey == witprog || scriptPubKey == GetScriptForDestination(p2sh)) {
+        assert(HaveKey(keyid));
+        // No need for a IsSolvable check here, as this function is only called for outputs
```

Comment:
Hmm, if I'm reading this correctly we technically could get here with an uncompressed keyid. Obviously this isn't the right place to check for that, but we should probably do *something* to handle it better (some big logprints, etc). Also need a test for this case, I'd think (make sure we dont consider it as valid balance, etc).

Reply:
If that happens, it's already too late. This function is only called for scriptPubKeys we're already considering ours.